### PR TITLE
1659 enable seller listing retirements UI

### LINKED ISSUE #1659: [seller listing] Enable RETIRE NOW functionality on a Seller listing
With CM Contract v2 and the return of Seller listings almost released in Production, the time is now right to add the capability to Retire from the listing on a Project page.

We have this functionality for Pool listings...
Screenshot 2023-10-13 at 9 28 46 AM

...but it was never added for Seller listings. 

Currently the way this is supported is that a user is required to BUY a carbon asset and then from their Portfolio page they can retire the carbon. The objective of this workstream is to support a RETIRE NOW option on seller listings; which would just be a combination of BUY & RETIRE.

There was a dependency for this to upgrade the retirement aggregator contract (https://github.com/KlimaDAO/klimadao-solidity/issues/48). This is ready to go for Toucan/C3 listings on the `infinity-carbonmark` branch (note: this can't be deployed until we have final contract addresses for carbonmark v2).

### Requirements

- [ ]  Add RETIRE NOW button to a Seller Listing card. The design UI should match what exists for a Pool listing (as per screenshot above).

- [ ] Click RETIRE NOW to open the /retire form, for [example](https://www.carbonmark.com/projects/VCS-1650-2015/retire/pools/nct). Note the UI of the form should match the UI as if the user clicked RETIRE NOW on a pool listing.

Screenshot 2023-10-16 at 9 20 29 AM

**NOTES:**
- will need the new ABI endpoint from the retirement aggregator contract
- will need to decide on the URL path of the form; assume it will be something like: https://www.carbonmark.com/projects/VCS-1650-2015/retire/seller/?

- [ ] As part of this change, note the Asset Details UI (closed by default)
- [ ] Note that in the Total price component, there are no network fees to buy or retire from a seller listing (nothing needs to change on the form, this is covered in the [docs](https://docs.carbonmark.com/get-started/understanding-fees-on-carbonmark))
- [ ] If paying with CC, payment processing fee still applies

**NOTES:**
- to support pay with CC, there is a dependency on changes required on the offsetra API; primarily around needing to support a listing ID from the ABI
- this is covered in a separate ticket #1818

### PR DESCRIPTION
**Pull request by sprrwhwk**
## Description

This PR contains the frontend changes to make retiring seller listings possible.

## Related Ticket

Closes #1659 

## Notes For QA
Pool purchases and retirements should also be tested.

### CODE REVIEW

**Review by Atmosfearful — DISMISSED** [review]
Super close, might be good to go once we fix the checkout api

File: carbonmark/pages/projects/[project_id]/retire/[listing_id].ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  try {
+    let project = await getProjectsId(project_id);
+
+    if (!project) {
+      throw new Error("No project found");
+    }
+
+    let listing = await getListingsId(listing_id.toLowerCase());
```
[the remark this answers is not in the document]
Comment (sprrwhwk) [reply carbonmark/pages/projects/[project_id]/retire/[listing_id].ts]: Good catch 👍 
Done in [7794810](https://github.com/KlimaDAO/klimadao/pull/2056/commits/77948108320c302c6ae66391f6f0a5379e75bcaf)

File: carbonmark/components/pages/Project/BuyOptions/SellerListing.tsx [hunk 2]
```diff
@@ -88,33 +93,24 @@ export const SellerListing: FC<Props> = (props) => {
...
-          className={styles.buyButton}
-          label={t({
-            id: "shared.connect_to_buy",
-            message: "Sign In / Connect To Buy",
-          })}
-          onClick={toggleModal}
+          label={t`Buy`}
+          renderLink={(linkProps) => <Link {...linkProps} />}
+          onClick={() => {
+            LO.track("Purchase - Listing: Buy Clicked");
```
Comment (sprrwhwk) [inline carbonmark/components/pages/Project/BuyOptions/SellerListing.tsx:101]: Not sure about these analytics events here. I just introduced a new event I guess. It's based on the pool equivalent

**Comment by jabby09** [comment]
@0xtapi this one is good to go

**Comment by 0xtapi** [comment]
This needs some more work. Caught these issues:

1. Retirements from seller listings don't go through, transaction can only be approved, but not submitted. No error message, the "Submit" button simply does nothing. Console says:
retireerror

2. There's something weird going on with the way prices are calculated. See this:
numbererr2

For some reason, the price increases by ~50% from "Price per tonne" to "Total cost". Also, when I enter a different number, like 2 tonnes, it quotes me the correct $2. When I go back to 1 tonne, I sometimes get asked for $1.67, sometimes $1.73... There's some wild math going on in the background. This appears to happen only if I try to retire 1 tonne. For all other tonne values, the asking price seems correct.

Note: if I retire/purchase from pools, this does not happen and all values seem to be correct.

3. Related to the issue above: The actual price here was $1.515, but the UI rounded it to $1.5. Also for prices, it should always show two decimals, imo. So show $1.50 if that's the actual price, never $1.5.

Short summary:
Retirements from seller listings don't work ❌
Weird math when retiring 1 tonne from sellers ❌
Some rounding errors during purchase process ❌

Pool purchases work ✅
Pool retiremens work ✅

File: carbonmark/components/pages/Project/Retire/Pool/RetireForm.tsx [hunk 3]
```diff
@@ -165,10 +175,17 @@ export const RetireForm: FC<Props> = (props) => {
...
-      project_address: !props.price.isPoolDefault
-        ? inputValues.projectTokenAddress
-        : null,
-      retirement_token: props.price.poolName.toLowerCase() as PoolToken,
+      project_address:
+        props.product.type === "pool"
+          ? !props.product.isPoolDefault
+            ? inputValues.projectTokenAddress
+            : null
+          : null,
```
Comment (0xemc) [inline carbonmark/components/pages/Project/Retire/Pool/RetireForm.tsx:183]:
```suggestion
isPool && isDefault ? inputValues.projectTokenAddress : null,
```

File: carbonmark/lib/constants.ts [hunk 4]
```diff
@@ -28,7 +28,7 @@ const SHORT_COMMIT_HASH = process.env.NEXT_PUBLIC_VERCEL_GIT_COMMIT_SHA?.slice(
 );
 
 /** When incrementing this API version, be sure to update TypeScript types to reflect API changes */
-export const API_PROD_URL = "https://v5.0.0.api.carbonmark.com";
+export const API_PROD_URL = "https://v5.1.0.api.carbonmark.com";
```
Comment (0xemc) [inline carbonmark/lib/constants.ts:31]: This might need an update?

**Review by 0xemc — APPROVED** [review]
Good to go once @0xtapi @jabby09 give this the tick

**Comment by 0xemc** [comment]
@sprrwhwk bump on this

File: carbonmark/components/pages/Project/Retire/Pool/RetireForm.tsx [hunk 5]
```diff
@@ -70,7 +70,10 @@ export const RetireForm: FC<Props> = (props) => {
   const methods = useForm<FormValues>({
     mode: "onChange",
     defaultValues: {
-      projectTokenAddress: props.price.projectTokenAddress,
+      projectTokenAddress:
+        props.product.type === "pool"
```
[the remark this answers is not in the document]
Comment (sprrwhwk) [reply carbonmark/components/pages/Project/Retire/Pool/RetireForm.tsx]: done in [a93b36d](https://github.com/KlimaDAO/klimadao/pull/2056/commits/a93b36d1672e270ec8d6b1392a7e3ce6020ea21a)

**Comment by 0xtapi** [comment]
Retiring from seller listing now works smoothly, asking price is also fixed 👍 

However, pool purchases seem to be broken. I did not check all 324 listings on CM, but I assume all of them are broken based on about a dozen projects I checked. Like these:

https://carbonmark-git-1659-enable-seller-listing-retir-5860b8-klimadao.vercel.app/projects/VCS-875-2015/purchase/pools/nbo
https://carbonmark-git-1659-enable-seller-listing-retir-5860b8-klimadao.vercel.app/projects/VCS-981-2012/purchase/pools/nct
https://carbonmark-git-1659-enable-seller-listing-retir-5860b8-klimadao.vercel.app/projects/VCS-191-2008/purchase/pools/bct

Result always looks like this:
pool

Purchasing from seller listings works, though. It's only the pools that are broken.

**Comment by 0xemc** [comment]
@sprrwhwk where are we at with this one?

**Comment by 0xemc** [comment]
@0xtapi if you can confirm this one again please then we can close it 🙏

**Comment by 0xtapi** [comment]
All good now, but two tiny issues remaining:
Screenshot 2024-01-25 115333

1. The link to PolygonScan currently points to Mumbai.
2. There's a space missing in the first link, "your[ ]retirement".

No need to test this again once this is fixed, could not find any problems. Looking great👍

**Pull request merged**